src/main/java/io/strimzi/kafka/quotas/CachingVolumeObserver.java

```diff
@@ -0,0 +1,154 @@
+ * can augment in observations from brokers that have dropped out of the results (due to being restarted or otherwise
+ * removed from the active set of brokers).
+ */
+public class CachingVolumeObserver implements VolumeObserver {
+    private final VolumeObserver observer;
+    private final Clock clock;
+    private final Duration entriesValidFor;
+    private final ConcurrentMap<CacheKey, VolumeUsage> cachedObservations;
+    private final LinkedHashMap<String, String> defaultTags;
+    private final Map<CacheKey, Counter> evictionsPerRemoteBroker = new ConcurrentHashMap<>();
+
+    private final Logger log = getLogger(CachingVolumeObserver.class);
+
+    /**
+     * @param observer The downstream observer to be notified after processing
+     * @param clock the clock to use for managing cache expiry
+     * @param entriesValidFor how long cache valid entries for
+     * @param defaultTags tags to be added to all metrics
+     */
+    public CachingVolumeObserver(VolumeObserver observer, Clock clock, Duration entriesValidFor, LinkedHashMap<String, String> defaultTags) {
+        this.observer = observer;
+        this.clock = clock;
+        this.entriesValidFor = entriesValidFor;
+        this.defaultTags = new LinkedHashMap<>(defaultTags);
+        cachedObservations = new ConcurrentHashMap<>();
+        Metrics.newGauge(metricName(CachingVolumeObserver.class, "CachedEntries", defaultTags), new Gauge<>() {
+            @Override
+            public Integer value() {
+                return cachedObservations.size();
+            }
+        });
+    }
+
+    @Override
+    public void observeVolumeUsage(VolumeUsageResult result) {
+        VolumeUsageResult outgoing;
+        if (result.getStatus() == SUCCESS) {
+            if (log.isDebugEnabled()) {
+                log.debug("Caching successful observation. Propagating observation along with cached values where appropriate to the next observer.");
+            }
+            outgoing = cacheAndAugment(result);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("Not caching failed observation. Propagating failed observation to the next observer.");
+            }
+            maybeExpireCachedObservations();
+            outgoing = result;
+        }
+        observer.observeVolumeUsage(outgoing);
+    }
+
+    private void maybeExpireCachedObservations() {
+        Set<CacheKey> toRemove = cachedObservations.entrySet().stream()
+                .filter(e -> isExpired(e.getValue())).map(Map.Entry::getKey).collect(Collectors.toSet());
+        toRemove.forEach(this::evict);
+    }
+
+    private void evict(CacheKey key) {
+        final Counter counter = getEvictionCounter(key);
+        counter.inc();
+        if (log.isDebugEnabled()) {
+            log.debug("evicting entry logDir: {} on broker: {}", key.logDir, key.brokerId);
+        }
+        cachedObservations.remove(key);
+    }
+
+    private Counter getEvictionCounter(CacheKey key) {
```

Comment:
The name of this method seems a bit misleading. The `get` suggests that it doesn't have side effects, but it's used on line 127 entirely for it's side-effect, where it looks a little odd that we're ignoring the return value. Perhaps naming this method `evictionCounter()` would be better?

Reply:
resolved in e49d2f21599add288c3285cf213a141432822fda